### ISSUE TITLE
RFC: Trusted Crates

### ISSUE THREAD

**Issue by bholley** [mozilla/cargo-vet#458]
This RFC proposes a mechanism for projects to trust certain authors to publish specific crates, as well as a mechanism for these trust decisions to be surfaced to other projects as hints.

# Motivation

The purpose of cargo-vet is to achieve supply-chain integrity by ensuring that the contents of each dependency are signed off by a trusted source. The tool is audit-oriented because the crates in the ecosystem are very heterogeneous in origin: it's usually impractical to require that every dependency was _developed_ by a trusted source, so the next best thing is to ensure that everything has been _audited_ by a trusted source.

However, there are some cases where the developer is in fact trusted. In these cases, an audit isn't really needed. If _all_ your dependencies were authored by a trusted source, you wouldn't bother to deploy cargo-vet in the first place. Put differently: if I trust you to faithfully audit a third-party crate and certify that it's safe-to-deploy, I also trust you to make the same assessment about code you wrote. This is why cargo-vet permits organizations to [self-certify](https://mozilla.github.io/cargo-vet/curating-your-audit-set.html#self-certification) crates which they develop and publish.

Self-certification has proven very useful and provided a lot of crate coverage. Unfortunately, it doesn't easily scale: the author needs to submit a certification for their crate to the cargo-vet instance of a widely-trusted organization, which is a very manual process. In practice, the primary impetus for generating self-certifications occurs when a developer attempts to bring their own externally-developed crate back into a project they're working on. So for example, @alexcrichton has certified audits in the [wasmtime audit set](https://github.com/bytecodealliance/wasmtime/blob/main/supply-chain/audits.toml) for a number of external crates which he personally developed and which wasmtime uses, but it's a fraction of the full set of [314 crates](https://crates.io/users/alexcrichton) he's published on crates.io. Moreover, many/most authors of widely-used crates do not have an obvious place to record their audits — we've [occasionally](https://searchfox.org/mozilla-central/rev/31f5847a4494b3646edabbdd7ea39cb88509afe2/supply-chain/audits.toml#1528) [dabbled](https://github.com/bytecodealliance/wasmtime/blob/a486aa37ad187e4b5c5d34da303a39fa5cdd2745/supply-chain/audits.toml#L1757) with "guest audits", but that approach is neither elegant nor efficient.

Ideally we would have a way for a consumer of a crate (and user of cargo-vet) to easily inspect its provenance, determine it's sufficiently trustworthy, and record this conclusion unilaterally. There's a delicate balancing act here between ergonomics/discoverability and security, which requires some care to resolve.

# Proposed Design

In `audits.toml`, we introduce a section called `trusted`. Entries look like this:

```
[[trusted.serde]]
criteria = "safe-to-deploy"
user-id = 3618 // David Tolnay
start date = ...
end-date = ...
notes = "David is super-trustworthy."
```

These behave similarly to [wildcard audits](https://mozilla.github.io/cargo-vet/wildcard-audit-entries.html), with two key differences:
* Semantically they do not constitute a certification, but rather a policy decision to trust crates published by non-participating individual. As a consequence, there is no "who" field.
* They do not propagate across imports.

Ordinarily non-importable things would go in `config.toml`. However, we shim them into `audits.toml` so that they can be used for _hinting_ during registry suggestions (see below).

When computing `suggest` results, we look up the publisher of every crate in the list. We list this publisher inline for every crate. If a crate is published by an author who appears in a `trusted` import by someone in the registry, we note it as a hint. Here's an example:

```
$ cargo vet suggest

recommended audits for safe-to-deploy:

cargo vet inspect pathdiff 0.2.1        by: Manishearth (used by open)                      (517 lines)
cargo vet inspect serde_json 1.0.82     by: dtolnay (used by insta, reqwest, and 3 others)  (22869 lines)
  NOTE: mozilla, bytecode-alliance, and fermyon trust dtolnay — consider cargo vet trust serde_json
```

A few notes:
* The suggestion behavior here is intended to apply cross-crate: if mozilla trusts dtonlay to publish `serde`, and the local project needs `serde_json`, the suggestion will still appear. The rationale is that an author that is trusted to publish crate `X` would _likely_ also be trusted to publish crate `Y` if it were in use. So we cast a wide net for suggestions to improve discoverability, but scope the actual trust entry to a specific crate/publisher/date-range to ensure an explicit decision on the actual trust delegation.
* If there’s a trusted entry which would apply _except_ that it’s outside the date range, we can add an explicit NOTE for that as well so that it’s easy to discover the issue and bump the range.
* Whereas a wildcard audit is an explicit promise from the publisher that the code will meet the criteria, this is a mere heuristic that may or may not hold in practice. In particular, the publisher may or may not have personally authored or reviewed all the code. Assessing the risk here requires some judgment and potentially some investigation (e.g., by examining the development activity in the upstream repository). The supply-chain owners have the opportunity to apply this judgment periodically when the expiration date is updated, or even for every version when the publisher information is reified in `imports.lock`.

**Comment by bholley** [comment]
CC @alexcrichton @pchickey @fitzgen @RobJellinghaus @repi @tschneidereit @divergentdave in case this is of interest.

**Comment by tschneidereit** [comment]
Thabk you for working through this and arriving at a very thoughtful design, Bobby! I agree with the reasoning behind the (meta-)policy decision here, and think the way of realizing it is very elegant.

**Comment by mystor** [comment]
This was implemented in #461
